reduction_scripts/reduce_data.py

```diff
@@ -0,0 +1,848 @@
+
+import sys
+import os
+import pickle
+from astropy.io import fits as pyfits
+import pywifes
+import gc
+import datetime
+import numpy as np
+from data_classifier import classify
+import json
+
+#------------------------------------------------------------------------
+start_time = datetime.datetime.now()
+#------------------------------------------------------------------------
+
+# Read the raw data directory from command line
+data_dir = os.path.abspath(sys.argv[1])+'/'
+
+
+# Classify all raw data (red and blue arm)
+obs_metadatas = classify(data_dir)
+
+
+for arm in obs_metadatas.keys():
+
+    obs_metadata = obs_metadatas[arm]
+    # Get the project directory from the file location 
+    proj_dir = os.path.dirname(__file__)
+    
+    # Check observing mode
+    sci_im = obs_metadatas[arm]['sci'][0]['sci'][0]+'.fits'
+    if pywifes.is_nodshuffle(data_dir+sci_im):
+        obs_mode = 'ns'
+    else:
+        obs_mode = 'class'    
+
+    #************************************************************************
+    #*****       LOAD JSON FILE WITH USER REDUCTION SETUP              *****
+    #************************************************************************
+    # Read the JSON file
+    file_path = 'params_'+obs_mode+'.json'
+    with open(file_path, 'r') as f:
+        proc_steps = json.load(f)
+    #************************************************************************
+
+
+    # Check if the reduc_blue/red folders already exists and create them if required
+    out_dir = os.path.join(proj_dir, 'reduc_'+arm) 
+    if not os.path.exists(out_dir):
+        os.mkdir(out_dir)
+    else:
+        print(f"Folder '{'reduc_'+arm}' already exists.")
+
+    
+    calib_prefix = os.path.join(out_dir, 'wifesR_20150314')
+
+    # Some WiFeS specific things
+    my_data_hdu=0
+
+    # SET MULTITHREAD ?
+    #~ multithread=False
+    multithread=False
+
+    # SET SKIP ALREADY DONE FILES ?
+    skip_done=False
+    #skip_done=True
+
+    #------------------------------------------------------------------------
+    #------------------------------------------------------------------------
+    # METADATA WRANGLING FUNCTION
+    def get_full_obs_list(metadata):
+        full_obs_list = []
+        base_fn_list = (
+            metadata['bias']+
+            metadata['arc']+
+            metadata['wire']+
+            #metadata['dark']+
+            metadata['domeflat']+
+            metadata['twiflat'])
+        for fn in base_fn_list:
+            if fn not in full_obs_list:
+                full_obs_list.append(fn)
+        for obs in (metadata['sci']+
+                    metadata['std']):
+            for key in obs.keys():
+                # Fred's update 2 (another consequence of it ...)
+                if (key != 'type') and (key != 'name'): 
+                    for fn in obs[key]:
+                        if fn not in full_obs_list:
+                            full_obs_list.append(fn)
+        return full_obs_list
```

Comment:
We could refactor this to define all the functions at the top of the file, then the loop could just parse metadata, load configs, and then run the process steps. 

But save it for a follow-up pull request. I think atm there's some global variables being defined in the loop, and it'd make the diff more complicated.

Reply:
Ok, good, I'll add it to the tasks list.